SARAALERT-1111: Refactor Reports Table into React Component

kylietmo opened the pull request:
# Description
Jira Ticket: [SARAALERT-1111](https://tracker.codev.mitre.org/browse/SARAALERT-1111)

This PR refactors the Patient reports/assessments table from a DataTable to a React component. The purpose for this change is so that we can finally paginate the assessments and gain a huge performance increase as jurisdictions such as GLITC are having trouble loading the Patient page due to many assessments. 

Visual changes:
- Actions button is now the leftmost column for easier access and is in a consistent style with the rest of the application.
- Input for entries per page is now below the table and looks more consistent with other tables in the application.
- Search bar is updated. 

Functional changes:
- Search bar only allows searching by ID or reporter. Before it was sort of a catch all for any text in the table.

NOTE: This also addresses this ticket about table overflow: [SARAALERT-1090](https://tracker.codev.mitre.org/browse/SARAALERT-1090). This was done by adding `responsive` to the `<Table>` element in `CustomTable`. 


# Demo/Screenshots
Before:
![Screen Shot 2020-12-18 at 12 52 14 PM](https://user-images.githubusercontent.com/11698457/102645063-dce20380-412f-11eb-872f-132bf9aab021.png)

After:
![Screen Shot 2020-12-18 at 12 42 10 PM](https://user-images.githubusercontent.com/11698457/102644236-74465700-412e-11eb-8969-32fb5fb3ef72.png)

# Important Changes
`app/views/patients/show.html.erb`
- Removed assessment DataTable and replaced with new React component.

`app/javascript/components/patient/PatientReportsTable.js`
`app/javascript/components/patient/ReportModal.js`
- New React components for table and related modal.

`app/controllers/assessments_controller.rb`
- New action to handle getting data for assessments table. Hit by new route.

`app/helpers/assessment_query_helper.rb`
- Helper methods for filtering/sorting/paginating/formatting assessments table data.

`app/javascript/components/layout/CustomTable.js`
- Clearer variable names
- Passes more params to filter method in colData (if provided)
- Now supports custom row className.

`app/javascript/components/admin/AdminTable.js`
- Updating filter method with new params. 
- Removing unnecessary code about tab,

`app/javascript/components/admin/AuditModal.js`
`app/javascript/components/public_health/PatientsTable.js`
- Updating filter method with new params. 

# Testing
This fix was tested on the following browsers (submitter must check all those that apply):
* [x] Chrome
* [ ] Firefox
* [ ] Safari
* [ ] IE11

**Please test the following:**
I think I enumerated most of it below, but please review the "Reports" section of the user guide [here](https://saraalert.org/wp-content/uploads/2020/12/Sara-Alert-User-Guide-v1.18.pdf) to understand the expected functionality. I found bugs with this PR by doing so. 
- [ ] All table functionality and buttons is the same as it was before.
    - [ ] Editing reports works
    - [ ] Adding reports works
    - [ ] Marking a report as reviewed works
    - [ ] Adding a note to a report works    
    - [ ] Marking all reports as reviewed works
    - [ ] "Mark All As Reviewed" and "Review" buttons are NOT available in isolation workflow
    - [ ] "Needs Review" column should NOT be shown in isolation workflow
    - [ ] In Exposure workflow, "Needs Review" column should have a tooltip next to it explaining the column
    - [ ] Rows are highlighted red when "needs review" column value is "Yes"
    - [ ] Text of values is highlighted when the value is "Yes" or is a numeric value that passes the threshold provided by the jurisdiction config.
    - [ ] When there is only a single assessment in the table, the actions dropdown should NOT be cut off
    - [ ] When response is received with SMS text and the response is "Yes" the row should be highlighted red by all the symptom col values should be blank
- [ ] Test that functionality is expected when looking a record that has been transferred between jurisdictions with different symptom configurations. Specifically:
    - [ ] Column list of symptoms should be AGGREGATE across all of the record's assessments. So if you transfer a record from USA, State 1 (with pulse ox) to USA, State 2 - it should still show the pulse ox column.
    - [ ] When adding a new record, should only show the current jurisdiction symptom options
    - [ ] When editing an assessment that was created when in a different jurisdiction, should show the options at the time for that assessment (so if it was created when in State 1, should show pulse ox option when editing)
- [ ] The table loads quickly with 500+ legitimate assessments (that have valid symptoms).
- [ ] The table can handle overflow.
- [ ] Admin table was not affected in any way visually/functionally
- [ ] Patient line-lists were not affected in any way visually/functionally

holmesie commented on app/controllers/assessments_controller.rb:
```diff
@@ -175,6 +176,31 @@ def update
     redirect_to(patient_assessments_url) && return
   end
 
+  def table
```
It would make more sense to call this `index`.

kylietmo replied:
Updated since changing request to a GET to use `index` now as well: https://github.com/SaraAlert/SaraAlert/pull/604/commits/30d4fbeb075c861ae8a66567d663f31a9b23ee50

holmesie commented on app/controllers/assessments_controller.rb:
```diff
@@ -175,6 +176,31 @@ def update
   end
 
+  def table
+    redirect_to(root_url) if ADMIN_OPTIONS['report_mode']
```
Is this really necessary, or just extra?

kylietmo replied:
I didn't want anyone to be able to hit this route at all (even if manually entering the url) so it was added out of an abundance of caution. I can remove if we feel it's not necessary at all.

holmesie replied:
Nope, just curious. I think it's fine to leave in!

holmesie commented on app/helpers/assessment_query_helper.rb:
```diff
@@ -5,4 +5,67 @@ module AssessmentQueryHelper
   end
+
+  def search(assessments, search)
+    return assessments if search.nil? || search.blank?
```
`search.nil? || search.blank?` can just be `search.blank?`

kylietmo replied:
Good catch. Fixed here: https://github.com/SaraAlert/SaraAlert/pull/604/commits/30d4fbeb075c861ae8a66567d663f31a9b23ee50

holmesie commented on app/javascript/components/admin/AdminTable.js:
```diff
@@ -639,12 +642,10 @@ class AdminTable extends React.Component {
                 }
                 className="ml-3"
                 disabled={!this.state.actionsEnabled}>
-                {this.state.query.tab !== 'closed' && (
```
This _looks_ like a good change, just a bit unexpected in this PR. Was this a bug?

kylietmo replied:
It was a remnant of when I first did the AdminTable refactor months ago it looks like. I was surprised to see it. The PatientsTable deals with tabs in the main query, but the AdminTable does not, so I feel it's a safe change. https://github.com/SaraAlert/SaraAlert/blob/7934c39add6ce917befceff91d2e11c21a17a8d7/app/javascript/components/public_health/PatientsTable.js#L78

holmesie replied:
Okay, got it

holmesie commented on app/javascript/components/admin/AuditModal.js:
```diff
@@ -126,9 +129,12 @@ class AuditModal extends React.Component {
+   * @param {Number} colData - Column data for the col this filter is a part of.
    */
-  formatChange = change => {
+  // eslint-disable-next-line no-unused-vars
```
Was eslint complaining about this?

kylietmo replied:
Yes, this function is called by the CustomTable filter which passes the same data to everyone but only some params are used by most. I'm likely going to add a change that packages it all into one object to avoid having to add this eslint line.

kylietmo replied:
This has since been removed: https://github.com/SaraAlert/SaraAlert/pull/604/commits/30d4fbeb075c861ae8a66567d663f31a9b23ee50

holmesie commented on app/javascript/components/layout/CustomTable.js:
```diff
@@ -138,7 +138,7 @@ class CustomTable extends React.Component {
           </div>
         )}
-        <Table striped bordered hover size="sm">
+        <Table responsive striped bordered hover size="sm">
```
Since all of our tables (that use this class) would inherit this, are we okay with this change globally?

kylietmo replied:
Yes, I believe this actually addresses the overflow issue described in this ticket: [SARAALERT-1090](https://tracker.codev.mitre.org/browse/SARAALERT-1090). Will do more testing to verify as well.

holmesie commented on app/javascript/components/patient/PatientReportsTable.js:
```diff
@@ -0,0 +1,372 @@
+  queryServer = _.debounce(query => {
+    axios.defaults.headers.common['X-CSRF-Token'] = this.props.authenticity_token;
+    axios
+      .post('/patients/' + this.props.patient.id + '/assessments/table', {
```
Does this one actually need to be a POST?

kylietmo replied:
I made it a POST To be consistent with the line list request, but given it's not really necessary and we want to use `index` for the route endpoint, I've updated it to be a GET here: https://github.com/SaraAlert/SaraAlert/pull/604/commits/30d4fbeb075c861ae8a66567d663f31a9b23ee50

holmesie commented on app/models/assessment.rb:
```diff
@@ -95,7 +95,7 @@ def get_reported_symptom_by_name(symptom_name)
     threshold_cond_hashes = ReportedCondition.where(type: 'ReportedCondition', assessment_id: assessment_ids).pluck(:threshold_condition_hash)
     condition_ids = ThresholdCondition.where(type: 'ThresholdCondition', threshold_condition_hash: threshold_cond_hashes)
-    Symptom.where(condition_id: condition_ids).pluck(:name)
+    Symptom.where(condition_id: condition_ids).pluck(:name).uniq.sort
```
Did you find `uniq` was necessary during testing?

kylietmo replied:
I included it for right now to be consistent with the existing functionality, but I haven't tested without just yet. My thinking is that it will be necessary, but not 100% sure yet.

kylietmo replied:
After some more testing and reviewing, the `uniq` is definitely necessary. Otherwise this will include all Symptoms for the threshold condition hashes for those assessments, and each assessment creates a set of new Symptoms, so there will be duplicates.

holmesie reviewed (changes requested):
Looking great! This was my first pass review, just have a few questions.

kylietmo commented on app/helpers/assessment_query_helper.rb:
```diff
@@ -5,4 +5,88 @@ module AssessmentQueryHelper
+      assessments = assessments.order(who_reported: dir)
+    when 'created_at'
+      assessments = assessments.order(created_at: dir)
+    else
```
I will be the first to say I do not like this implementation for sorting the symptom columns. It is slow, it requires the use of a new gem, and it's a bit hard to follow. I feel like there _must_ be a better way to handle this sorting and would be happy to brainstorm more, but none of my attempts with a joins/includes worked out in such a way that I could get this order in a Relation by the end.

kylietmo commented on app/controllers/assessments_controller.rb:
```diff
@@ -131,7 +156,6 @@ def create
         # Create history if assessment was created by user
         History.report_created(patient: patient, created_by: current_user.email, comment: "User created a new report (ID: #{@assessment.id}).") if current_user
       end
-      redirect_to(patient_assessments_url)
```
@holmesie I removed this because before it didn't seem to actually be doing anything (since `index` was empty) and now it is causing issues with it defined. Was there a reason for it before?

holmesie replied:
The front end doesn't work this way anymore for assessments, so this really isn't needed.

kylietmo commented on app/javascript/components/layout/CustomTable.js:
```diff
@@ -138,62 +138,71 @@ class CustomTable extends React.Component {
-                        {col.onClick && <span onClick={() => (col.onClick(data.id.toString()) ? col.onClick : null)}>{value}</span>}
-                        {!col.onClick && value}
+        <div></div>
+        <div className={this.props.getCustomTableClassName ? `table-responsive ${this.props.getCustomTableClassName()}` : 'table-responsive'}>
```
FYI: This diff is large mostly because I added a div around the Table element for responsiveness and custom classes.

Other changes included that are now not so obvious (thanks GitHub):

- Updating variable names in the row/col loops.
- Passing more data to the provided col filter method (if any).
- New prop for row className.

-

kylietmo commented on app/controllers/assessments_controller.rb:
```diff
@@ -101,6 +126,8 @@ def create
       redirect_to(root_url) && return unless patient
 
       threshold_condition_hash = params.permit(:threshold_hash)[:threshold_hash]
+      redirect_to(root_url) && return if threshold_condition_hash.blank?
```
@holmesie I added this redirect in the `create` because I found a bug in this PR where I could add assessments that would show up fine in the table, but actually had ReportedConditions with a nil threshold_condition_hash because for some reason, there is always a single ThresholdCondition with a nil hash in every environment so the redirect below this one was not being hit. Do you know why that single ThresholdCondition with a nil hash seems to exist?

kylietmo replied:
Update after speaking with @mayerm94. There are more than one instance of ThresholdConditions with `nil` hashes, and these are the threshold conditions at the individual jurisdiction level. The `ThresholdConditions` with a hash represents the condition that is the sum of all the threshold conditions (without hashes) throughout the jurisdiction path === state 1s custom symptoms + all of USAs symptoms, so the ones with `nil` hashes will be aggregated up. No assessments should point to a threshold condition with a nil hash, so this redirect should be safe.

kylietmo commented on app/javascript/components/assessment/Assessment.js:
```diff
@@ -62,17 +62,8 @@ class Assessment extends React.Component {
 
-  // TODO: This needs to use generic symptoms lists and not be hard-coded
   hasChanges() {
-    let currentAssessment = _.cloneDeep(this.props.assessment);
```
@holmesie @DPC15038 I removed this code because it did not appear to actually be doing anything (since `currentAssessment[parseInt(symptom)]` was undefined always it seems) and it had hardcoded values anyway that would have been wrong if it _was_ working. Let me know if it does need to be there for any reason. Tagged David since it looks like he added this code originally months ago.

holmesie replied:
I think this might have actually been @mayerm94

hackrm commented:
Noticed a minor issue with the pagination - When you select a new number of entries to display in the table, the pagination doesn't default back to page 1. So in the case where a user is on a page that wouldn't exist any longer due to the pagination update, you can get to this state where it doesn't look like the user has a particular page selected and the table blanks out. In the case of the screenshot, I was originally on page 35 which didn't exist any longer after I updated the number of entries displayed on a given page. 

![image](https://user-images.githubusercontent.com/42656458/102794765-d646cc80-4379-11eb-8e84-ad1381e93555.png)

kylietmo commented:
> Noticed a minor issue with the pagination - When you select a new number of entries to display in the table, the pagination doesn't default back to page 1. So in the case where a user is on a page that wouldn't exist any longer due to the pagination update, you can get to this state where it doesn't look like the user has a particular page selected and the table blanks out. In the case of the screenshot, I was originally on page 35 which didn't exist any longer after I updated the number of entries displayed on a given page.

Great catch @hackrm! I fixed this for PatientReportsTable, AdminTable, and the AuditModal (since they all had this issue). PatientsTable seemed to already reset the page to 0 in this case. See change here: 1fc4aeb

hackrm reviewed (approved):
Looks good to me!
